explorer/interpreter/type_checker.cpp

```diff
@@ -4556,6 +4567,22 @@ auto TypeChecker::DeclareClassDeclaration(Nonnull<ClassDeclaration*> class_decl,
     class_vtable[fun->name()] = {fun, class_level};
   }
 
+  if (class_decl->extensibility() == ClassExtensibility::None) {
+    for (const auto& vt : class_vtable) {
+      const auto* const fun = vt.getValue().first;
+      if (!fun->is_method()) {
+        continue;
+      }
+
+      if (fun->virt_override() == VirtualOverride::Abstract) {
+        return ProgramError(class_decl->source_loc())
+               << "Error declaring `" << class_decl->name() << "`"
+               << ": final class should implement abstract method `"
+               << fun->name() << "`.";
+      }
+    }
+  }
+
```

Comment:
A suggestion of check that you could reuse for variable declaration:

```suggestion
    CARBON_CHECK(CheckVTableHasNoAbstractMethods(class_vtable)) << 
               "Error declaring `" << class_decl->name() << "`"
               << ": `final` class should implement abstract method `"
               << fun->name() << "`.";
  }
```